### DISCUSSION TITLE
Save and restore chats

### DISCUSSION THREAD

**Discussion by soundweaverz** [olimorris/codecompanion.nvim#139] [General]
Hi Oli! Amazing work! I've been a Neovim user for a few years, shortly tried Zed, but switched back now. However, one feature I'm missing is the ability to save and restore chats and IIRC there used to be a "save and restore" bullet point in the README. I also found another discussion, in which you mention that it could be possible. Any idea how this is prioritized? If I can help you setting it up, I'd gladly try to help working it out!

**Comment by olimorris** [comment]
Hey @soundweaverz and thanks for your kind words.

I had them in the plugin up until a few days ago (9c57c35). Unfortunately the way I rewrote the messaging element of the chat buffer broke them. I deemed that it wasn't worthwhile fixing them as the future changes I have will necessitate a rewrite again, in a week or two.

Feel free to look through the commit and if you can think of a good way of saving and restoring chats, let me know. I'm open to a PR to implement them again.

**Reply by olimorris** [reply]
@soundweaverz I meant to say, that you can have multiple chat buffers open at the same time, and can then navigate between them with `}` and `{`. Not sure if that does away with your need to save chats but thought I'd mention it.

**Reply by a-priestley** [reply]
Are there any updates on this topic? I'll give my +1 to request restoring this functionality -- I personally find it very important

**Reply by olimorris** [reply]
No plans to add this from my end. As per my comment above, I'm open to a PR.

I presume to save chats now, you're copying the chat buffer and saving it down to a file?

**Reply by a-priestley** [reply]
Not currently. Right now I'm just leaving my neovim session open until I'm sure I won't need my running chats any longer. I no longer use `<C-c>` to close the chat and instead use `:CodeCompanionToggle` to make sure the chats aren't killed.

How did saving and loading work previously? If I save a chat buffer to a file, create a new chat, copy the old one to the buffer and submit a new prompt, will the running model be provided with the context it needs, as if I never closed the chat in the first place?

Edit:

I just realized I can do:
```lua
close = {
	modes = {
		n = "<C-c>",
		i = "<C-c>",
	},
	index = 3,
	callback = function()
		require("codecompanion").toggle()
	end,
	description = "Toggle Chat",
},
```

To override the default close functionality. Very handy!

**Reply by olimorris** [reply]
And you can also have multiple chat buffers open at the same time and cycle through them with `}` and `{`.

Previously the whole of the messages would be exported to a JSON file along with the adapter. If they're not in the `Chat.messages` table then they won't be sent to the LLM with `Chat:submit()`. I've included the commit which had the `saved_chats` functionality. above.

**Comment by itsfrank** [comment]
Howdy! I implemented a save/retrieve functionality I am pretty happy with you can find it in this gist: https://gist.github.com/itsfrank/942780f88472a14c9cbb3169012a3328

Pretty simple, it adds 2 commands:
- `CodeCompanionSave with some args` will save the current chat as a file whose name is the concatenation of the args with '-'
  - the incantation above would save it as `with-some-args.md`
- `CodeCompanionLoad` will open a telescope picker showing you all saved chats
  - the `<c-d>` mapping will delete currently selected chat

Works very well for my needs. The one thing it does not do is allow you to resume chatting, but I haven't really had the need for that. In any case I'm pretty sure I could just select the bits that are relevant and use `CodeCompanionAdd` to start a new chat with the relevant context

**Reply by olimorris** [reply]
Thanks for sharing! This is great.

In terms of resuming chatting...you could open up a chat buffer and pre-fill the messages table. It's a little more involved but something to consider.

**Reply by itsfrank** [reply]
That would definitely work, if I find myself needing this functionality a few times I'll update the gist!

**Reply by jemorriso** [reply]
hey, this is awesome! thanks for sharing, it's super helpful for me

**Reply by fredrikaverpil** [reply]
@olimorris It would be super useful to save/restore and _resume_ chats, preferably automatically and seamlessly. Is this anything you have on the roadmap?

I frequently find myself doing work with codecompanion where I have to restart Neovim (during nvim/lua plugin development) and would've loved to just keep on going with codecompanion.

Using `CodeCompanionSave` and `CodeCompanionLoad` requires quite the dance and is far from a good and seamless workflow here.

**Comment by kelvinauta** [comment]
+1 I really believe that saving chats is a very important functionality, as it is quite uncomfortable to rely on the session's integrity for chats, especially because it requires an effort to provide a chat with different contexts

**Comment by olimorris** [comment]
Hi All,

Wanted to share my position regarding this as I still have no plans to support this feature natively in the plugin. The primary reason is supporting saving/restoring will eat into the time I have to add new features. I'm speaking from experience after removing this feature back in August. I believe a native implementation tightly couples this feature to the chat buffer...something which changes regularly! This would mean breaking changes regularly and your saved chats either being unable to be restored or I'd need to come up with some bridging mechanism.

I believe we're reaching a point in the plugin's development where you can hook into functionality yourself in order to customize it to your workflow needs. This is something I've implemented in [another plugin](https://github.com/olimorris/persisted.nvim?tab=readme-ov-file#building_construction-extending-the-plugin) of mine recently where users were raising awesome ideas that were just too specific to their use cases.

In the case of saving and restoring, you could add your own custom keymap that implements the great work done in https://gist.github.com/itsfrank/942780f88472a14c9cbb3169012a3328.

Whilst I won't be adding this feature natively, if there are any hooks/events that I can add into the plugin, please let me know.

**Reply by fredrikaverpil** [reply]
I totally respect your stance on this. 👍 

I've been trying to use the code from the gist and make it into something that actually makes up a good workflow. I think the dealbreaker for me is you cannot load up a saved chat and resume chatting in the buffer (without copying its contents into a fresh codecompanion instance). Do you know what it might take to hook codecompanion back up with a loaded chat buffer?

**Reply by olimorris** [reply]
Assuming you saved file on disk is just the markdown content, you could create a shortcut that opens up a new chat buffer and pastes the content in. If you've used any references, that could be an issue as they'd need to be manually added back and potentially removed before they're saved to disk.

**Reply by fredrikaverpil** [reply]
I think I have something working, in case anyone else is interested:

- [codecompanion.lua](https://github.com/fredrikaverpil/dotfiles/blob/64d47392e4e59684a1181e8d957a2433b04625c5/nvim-fredrik/lua/fredrik/plugins/codecompanion.lua)
- [keymaps.lua](https://github.com/fredrikaverpil/dotfiles/blob/64d47392e4e59684a1181e8d957a2433b04625c5/nvim-fredrik/lua/fredrik/config/keymaps.lua#L1158#L1140-L1161)

It will use `CodeCompanionLoad` and ask you for the session you want to resume chatting with. Then it will ask you for the desired adapter. Finally it will copy paste the contents into a new codecompanion chat buffer. You need to make sure you pick the same adapter you used for the original chat, or you'll likely see an error. A tip is to enable `display.chat.show_setting = true`, which will add the model to the chat files.

EDIT: updated for fzf-lua.

**Reply by djmasaka** [reply]
Thank you @fredrikaverpil I used your code also when I pulled up the chat with a different adapter it still worked (and I could continue the chat) although I haven't tested beyond that.

**Reply by ChristinWhite** [reply]
Would an automated 'save chat history as Markdown files' be something more in the scope of what you're trying to do? While save and restore would be nice simply having my history automatically saved to my Obsidian vault would be a huge improvement since I constantly forget to copy it over manually.

**Comment by olimorris** [comment]
Hi All - As of today, we now have [codecompanion-history.nvim](https://github.com/ravitemer/codecompanion-history.nvim). You can read about installing it as an extension to CodeCompanion, [here](https://codecompanion.olimorris.dev/extensions/history.html).

**Reply by Davey-Hughes** [reply]
Hi, this plugin is amazing, and I completely support that it's an extension rather than built in- however it was a bit annoying that I had to come find this thread to discover the extension, given how natural it is to use history with LLMs. The link to the documentation is broken, and I'm not sure if there's currently a documentation overhaul going on because in learning to use codecompanion, a lot of the old documentation links from various github issues are broken.

Happy to contribute to docs if there's a need but just wanted to let you know that it was probably harder than it needed to be to find this.
